feat(peer-exchange): support peer retrieval by cluster/shard (#628)
vacp2p/rfc: content/docs/rfcs/34/README.md

This both protects the responder's anonymity as well as helps distributing load.

To allow for fast responses, responders SHOULD retrieve peers unsolicited (before receiving a query)

> and maintain a queue of peers by their `cluster_id`/`shard`  for the purpose of providing them in peer exchange responses.

Reviewer: Might need to link to the RFC here where these concepts are defined.

Author: There is no rfc for this. Previous approve was to maintain a LRU for peers, and remove the oldest when the capacity is reached. This cache data structure should account for cluster id and shard id for easier retrieval by shard. This is the PR where i created the maintained ds for go-waku. https://github.com/waku-org/go-waku/pull/840

To get the best anonymity properties with respect to response peer sets, responders SHOULD use each of these peers only once.

To save bandwidth, and as a trade off to anonymity,
responders MAY maintain a larger cache of exchange peers and randomly sample response sets from this local cache.

> This local cache SHOULD store peers by `cluster_id`/`shard` for faster retrieval. For peers without shard info, cache can maintain another list of peers.

Reviewer: ```suggestion
This local cache SHOULD store peers by `cluster_id`/`shard` for faster retrieval.
For peers without shard info, the cache can maintain a separate set of peers.
```

Semantic breaks = new line for new sentences.

> The size of each `cluster_id`/`shard` pair being maintained in cache SHOULD be large enough to allow randomly sampling peer sets that (on average) do not overlap too much.

Reviewer: ```suggestion
The size of the cache for each `cluster_id`/`shard` pair SHOULD be large enough to allow randomly sampling peer sets that (on average) do not overlap too much.
```

The `enr` field contains a Waku ENR as specified in [31/WAKU2-ENR](/spec/31/).

Requesters send a `PeerExchangeQuery` to a peer.

> When `cluster_id` is set and `shards` are specified, responder SHOULD provide peers belonging to cluster_id/shards pair.

Reviewer: ```suggestion
When `cluster_id` is set and `shards` are specified, responders SHOULD provide peers belonging to the specified cluster_id/shards pair.
```
Full sentences for RFCs, in other words it would be either `the responder` or `responders`.

> When no `shards` are specified, responder CAN provide mixed list of peers from any shard and non-shard speific peers. 

Reviewer: ```suggestion
When no `shards` are specified, responders SHOULD provide a mixed list of peers from any shard and non-shard specific peers. 
```
`CAN` is not a recognised RFC keyword: https://www.ietf.org/rfc/rfc2119.txt and in such cases you should use `MAY`. In this case, though, I would say that the relationship is `SHOULD`.

depends on the average number of requested peers, which is expected to be the outbound degree of the underlying
[libp2p gossipsub](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md) mesh network.
The recommended value for this outbound degree is 6 (see parameter `D` in [29/WAKU2-CONFIG](/spec/29/)).

> It is recommended for the cache to hold at least 10 times as many peers (60) for each `cluster_id`/`shard` pair. Size of whole cache is bounded by `no of cluster_id/shard pair` * `D` * 10.

Reviewer: ```suggestion
It is recommended for the cache to hold at least 10 times as many peers (60) for each `cluster_id`/`shard` pair. 
The size of the whole cache is bounded by `No. of cluster_id/shard pair` * `D` * 10.
```
